p256/src/arithmetic/field/field_risc0.rs

```diff
@@ -0,0 +1,101 @@
+//! 64-bit secp256r1 field element algorithms.
+
+use super::{MODULUS, MODULUS_HEX};
+use elliptic_curve::bigint::{risc0, Limb, U256};
+
+const MODULUS_256: U256 = U256::from_be_hex(MODULUS_HEX);
+
+pub(super) const fn add(a: U256, b: U256) -> U256 {
+    let a = a.as_limbs();
+    let b = b.as_limbs();
+
+    // Bit 256 of p is set, so addition can result in nine words.
+    // let (w0, carry) = adc(a[0], b[0], 0);
+    let (w0, carry) = a[0].adc(b[0], Limb::ZERO);
+    let (w1, carry) = a[1].adc(b[1], carry);
+    let (w2, carry) = a[2].adc(b[2], carry);
+    let (w3, carry) = a[3].adc(b[3], carry);
+    let (w4, carry) = a[4].adc(b[4], carry);
+    let (w5, carry) = a[5].adc(b[5], carry);
+    let (w6, carry) = a[6].adc(b[6], carry);
+    let (w7, w8) = a[7].adc(b[7], carry);
+    // Attempt to subtract the modulus, to ensure the result is in the field.
+    let modulus = MODULUS.0.as_limbs();
+
+    let (result, _) = sub_inner(
```

Comment:
If this `add` is run on denormalized inputs, it can result in incorrect outputs. For example, if `a = b = 2^256-1` then when you subtract off the modulus in `sub_inner` you are left with something a little larger than 2^256. This means the final `borrow` in the `sub_inner` algorithm will be 0x0000...01, which will then get `bitand`ed with each limb, which will clearly not be the correct answer.

With an honest host this shouldn't come up (because they'd just use normalized values), but a dishonest host could use denormalized values to "prove" a calculation that actually produces this incorrect answer.

Reply:
I observed that in the k256 implementation we have a `normalize()` function that checks if the value is fully normalized. What I have done here is to implement the same function and check if `a` and `b` are fully normalized first before performing the addition, to avoid the case of a dishonest host. Will adding this additional check help instead?

Reply:
If you check that these input values are normalized (or know by construction that they are normalized) then the algorithm you have here is safe. So yes adding the normalization check addresses this.